cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/InMemoryProgramRunDispatcher.java

```diff
@@ -0,0 +1,452 @@
+
+/**
+ * In-memory Implementation for {@link ProgramRunDispatcher} which runs the program in the same service from which it's
+ * invoked.
+ */
+public class InMemoryProgramRunDispatcher implements ProgramRunDispatcher, Closeable {
+
+  private static final String CLUSTER_SCOPE = "cluster";
+  private static final String APPLICATION_SCOPE = "app";
+  private static final Logger LOG = LoggerFactory.getLogger(InMemoryProgramRunDispatcher.class);
+
+  private final CConfiguration cConf;
+  private final Impersonator impersonator;
+  private final ArtifactRepository artifactRepository;
+  private final ProgramRunnerFactory programRunnerFactory;
+  private final ConfiguratorFactory configuratorFactory;
+  private final ExecutorService executorService;
+  private ProgramRunnerFactory remoteProgramRunnerFactory;
+  private String hostname;
+
+  @Inject
+  public InMemoryProgramRunDispatcher(CConfiguration cConf, ProgramRunnerFactory programRunnerFactory,
+      ConfiguratorFactory configuratorFactory, Impersonator impersonator,
+      @Named(AppFabricServiceRuntimeModule.NOAUTH_ARTIFACT_REPO) ArtifactRepository artifactRepository) {
+    this.cConf = cConf;
+    this.programRunnerFactory = programRunnerFactory;
+    this.configuratorFactory = configuratorFactory;
+    this.impersonator = impersonator;
+    this.artifactRepository = artifactRepository;
+    this.executorService = Executors.newFixedThreadPool(cConf.getInt(Constants.AppFabric.PROGRAM_CLEANUP_THREADS, 1));
+  }
+
+  /**
+   * Optional guice injection for the {@link ProgramRunnerFactory} used for remote execution. It is optional because in
+   * unit-test we don't have need for that.
+   */
+  @Inject(optional = true)
+  public void setRemoteProgramRunnerFactory(@Constants.AppFabric.RemoteExecution ProgramRunnerFactory runnerFactory) {
+    this.remoteProgramRunnerFactory = runnerFactory;
+  }
+
+  @Inject(optional = true)
+  public void setHostname(@Named(Constants.Service.MASTER_SERVICES_BIND_ADDRESS) InetAddress host) {
+    if (Objects.nonNull(host)) {
+      this.hostname = host.getCanonicalHostName();
+    }
+  }
+
+  @Override
+  public ProgramController dispatchProgram(ProgramRunDispatcherInfo programRunDispatcherInfo) throws Exception {
+    RunId runId = programRunDispatcherInfo.getRunId();
+    LOG.debug("Preparing to dispatch program run: {}", runId);
+    ProgramDescriptor programDescriptor = programRunDispatcherInfo.getProgramDescriptor();
+    ProgramOptions options = programRunDispatcherInfo.getProgramOptions();
+    boolean isDistributed = programRunDispatcherInfo.isDistributed();
+    ProgramId programId = programDescriptor.getProgramId();
+    ClusterMode clusterMode = ProgramRunners.getClusterMode(options);
+
+    // Creates the ProgramRunner based on the cluster mode
+    ProgramRunner runner = (clusterMode == ClusterMode.ON_PREMISE ? programRunnerFactory
+        : Optional.ofNullable(remoteProgramRunnerFactory).orElseThrow(UnsupportedOperationException::new)).create(
```

Comment:
indentation should be 2 spaces, looks like IDE still has something off when there is a newline.

Reply:
Have made changes to the setting. Should be ok now.